### PR TITLE
added separate permission for geocoder

### PR DESCRIPTION
**Pull request by adil-uddin**
https://dimagi-dev.atlassian.net/browse/USH-212

##### RISK ASSESSMENT / QA PLAN
[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-2081)

### CODE REVIEW

File: corehq/apps/accounting/bootstrap/features.py [hunk 1]
```diff
@@ -85,6 +85,7 @@
     privileges.TEMPLATED_INTENTS,
     privileges.RESTRICT_ACCESS_BY_LOCATION,
     privileges.REPORT_BUILDER_5,
+    privileges.GEOCODER,
```
Comment (orangejenny) [inline corehq/apps/accounting/bootstrap/features.py:88]: @biyeun How do we handle grandfathered plans when adding a new permission?

File: corehq/apps/hqadmin/management/commands/cchq_prbac_bootstrap.py [hunk 2]
```diff
@@ -165,6 +165,7 @@ def ensure_roles(self, roles, dry_run=False):
         Role(slug=privileges.APP_USER_PROFILES,
              name='App User Profiles',
              description='Allows use of App User Profiles.'),
+        Role(slug=privileges.GEOCODER, name="Project's default location.", description=''),
```
Comment (orangejenny) [inline corehq/apps/hqadmin/management/commands/cchq_prbac_bootstrap.py:168]: This and the other user-facing text should be something like "Address widget in Web Apps" although perhaps Kat or Dev will have a better suggestion on the ticket.

Comment (adil-uddin) [reply corehq/apps/hqadmin/management/commands/cchq_prbac_bootstrap.py]: Added name according to Kat
https://dimagi-dev.atlassian.net/browse/USH-212?focusedCommentId=93576

File: corehq/apps/accounting/bootstrap/features.py (hunk 1 above)
Comment (orangejenny) [reply corehq/apps/accounting/bootstrap/features.py]: Nevermind, resolved.

**Comment by orangejenny** [comment]
The existing code looks good. There's another step that I hadn't thought about previously: right now, the privilege only controls access to the default location setting, but it should also control access to the widget itself. The way I'd suggest doing this is checking in the django view, then passing a flag into Web Apps.

We have similar logic for handling feature flags: [here](https://github.com/dimagi/commcare-hq/blob/8a9a45ef44f65e7fa5db7f18992a22bc287ee67f/corehq/apps/cloudcare/views.py#L219) is a view checking for a feature flag, which is then passed along to javascript using [initial page data](https://github.com/dimagi/commcare-hq/blob/master/docs/js-guide/integration-patterns.md#javascript-in-django-templates). Then you'll want to check the flag when the question is created, around [here](https://github.com/dimagi/commcare-hq/blob/8a9a45ef44f65e7fa5db7f18992a22bc287ee67f/corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js#L976).

**Comment by orangejenny** [comment]
Thought of one more thing: typically when we add a new privilege, we grandfather in existing plans, so that no one who has already been using the feature loses access to it. The `cchq_prbac_grandfather_privs` does this. For an example, @stephherbers used it [here](https://github.com/dimagi/commcare-hq/pull/28543/files#diff-a4e7a48bc74f85c0212a722074b38f0b1866586c5f8e08cc0ded8759e2a03034) when she added a new privilege for app user profiles. Can you add a similar migration here?

File: corehq/apps/domain/migrations/0008_geocoder.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+from corehq.privileges import GEOCODER
+
+
+@skip_on_fresh_install
+def _grandfather_basic_privs(apps, schema_editor):
+    call_command('cchq_prbac_bootstrap')
+    call_command(
+        'cchq_prbac_grandfather_privs',
+        GEOCODER,
+        skip_edition='Paused,Community,Standard,Pro,Advanced',
```
Comment (orangejenny) [inline corehq/apps/domain/migrations/0008_geocoder.py:14]: I don't think this should skip advanced plans. I don't know the accounting code very well, that comment is based on looking at https://github.com/dimagi/commcare-hq/commit/b22aba39401af264282acc8f140894ce8cd1206f

File: corehq/apps/domain/migrations/0008_geocoder.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        noinput=True,
+    )
+
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ('domain', '0007_auto_20200924_1753'),
```
Comment (orangejenny) [inline corehq/apps/domain/migrations/0008_geocoder.py:21]: Can you move this into accounting, for the sake of consistency? All other migrations that call `cchq_prbac_grandfather_privs` are in accounting.

**Review by orangejenny — COMMENTED** [review]
A couple more minor comments, but this is looking good.

File: corehq/apps/accounting/migrations/0051_geocoder.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        'cchq_prbac_grandfather_privs',
+        GEOCODER,
+        skip_edition='Paused,Community,Standard,Pro',
+        noinput=True,
+    )
+
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ('accounting', '0049_app_user_profiles'),
```
Comment (biyeun) [inline corehq/apps/accounting/migrations/0051_geocoder.py:21]: the name of the migration dependency looks incorrect

**Review by biyeun — CHANGES_REQUESTED** [review]

**Review by orangejenny — APPROVED** [review]
Looks good to me. @biyeun I believe your comments are addressed, can you take a look?

**Review by biyeun — APPROVED** [review]
all set! 🎉

File: corehq/apps/cloudcare/templates/cloudcare/formplayer_home.html [hunk 6]
```diff
@@ -79,6 +79,7 @@
   {% initial_page_data 'single_app_mode' single_app_mode %}
   {% initial_page_data 'username' username %}
   {% initial_page_data 'use_live_query' use_live_query %}
+  {% initial_page_data 'has_geocoder_privs' has_geocoder_privs %}
```
Comment (orangejenny) [inline corehq/apps/cloudcare/templates/cloudcare/formplayer_home.html:82]: Have you tested in both web apps and app preview? I think app preview uses a different base template.

**Review by orangejenny — APPROVED** [review]
This looks good. I'm going to hold off on merging until after today's deploy.

**Review by orangejenny — APPROVED** [review]

**Pull request merged**